### ISSUE TITLE
io_uring_wait_cqe_nr() waits for wait_nr - 1 instead of wait_nr

### ISSUE THREAD

**Issue by marcelo-gonzalez** [axboe/liburing#257]
Hi!

Looking at the _io_uring_get_cqe() function, I noticed that there's this logic in there:
```
if (data->wait_nr && cqe)
	data->wait_nr--;
```

This makes it so that if you call io_uring_wait_cqe_nr(), and there's already at least one cqe available at the time of the call, then you'll return from that function after waiting for 1 less than you wanted. I think this could be fixed by just deleting those two lines if I understand correctly? Thought I would point it out, but I didnt want to send a patch because I don't know if it's safe to fix it, since someone who has a bug in their code and does this:
```
for (int i = 0; i < 4; i++)
  add_sqe(&ring, sqe);
do_something_that_takes_time();
io_uring_wait_cqe_nr(&ring, &cqe, 5);
```

would be broken by fixing this. Up to you!

thanks,

-Marcelo

**Comment by isilence** [comment]
It's not the only scenario how it can happen, e.g. `LIBURING_UDATA_TIMEOUT`, and there are other subtle problems. I'd suggest:

1. kill `cqe_ptr` parameter from `_io_uring_get_cqe()`, as well as `cqe` and `__io_uring_peek_cqe()` call in it.
2. simplify wait checks to something like
```C
need_enter = false;
const int wait_nr = data->wait;
...
if (wait_nr != io_uring_cq_ready())
    need_enter = true;
...
```
3. There is a subtle difference between iopoll and not. When `nr_wait` passed to non-IOPOLL, it wait for such number of requests to be in the ring, including ones already standing in there. However, IOPOLL will _try_ to get `nr_wait` new CQEs regardless of how much was in the CQ before. I need to check but AFAIR. As that was as such for long, IMHO better to work around in liburing

```C
int real_nr_wait = nr_wait;
if (IOPOLL)
    real_nr_wait -= io_uring_cq_ready();
```

4. for all exported functions returning an CQE (i.e. `**cqe_ptr`) get it separately. e.g. see `io_uring_for_each_cqe`.
5. if not used remove back `nr_available` from `__io_uring_peek_cqe` if not used.
6. ... double check to not screw other parts, like LIBURING_UDATA_TIMEOUT, `sq_ring_needs_enter()`, etc.

**Comment by marcelo-gonzalez** [comment]
hm yeah maybe that API would make sense. Feel free to just close this if you're going to implement it that way, since you're right, lots of other subtle stuff can happen. 

As far as the IOPOLL, i think it returns immediately if there are any cqes at all, if I understand right?

In fs/io_uring.c, io_iopoll_check():

```
                 * Don't enter poll loop if we already have events pending.
                 * If we do, we can potentially be spinning for commands that
                 * already triggered a CQE (eg in error).
                 */
                if (io_cqring_events(ctx, false))
                        break;

```
so I guess either way if io_uring_cq_ready() is > 0 it won't wait

**Comment by isilence** [comment]
> hm yeah maybe that API would make sense. Feel free to just close this if you're going to implement it that way, since you're right, lots of other subtle stuff can happen.

Feel free to take on it if you want, it may be a pretty useful thing to do!

> 
> As far as the IOPOLL, i think it returns immediately if there are any cqes at all, if I understand right?

Right, then that iopoll adjustment isn't needed.

> 
> In fs/io_uring.c, io_iopoll_check():
> 
> ```
>                  * Don't enter poll loop if we already have events pending.
>                  * If we do, we can potentially be spinning for commands that
>                  * already triggered a CQE (eg in error).
>                  */
>                 if (io_cqring_events(ctx, false))
>                         break;
> ```
> 
> so I guess either way if io_uring_cq_ready() is > 0 it won't wait

**Comment by isilence** [comment]
@marcelo-gonzalez, I'm curious if you're going to patch it? no pushing.
There are also sweet related spots for making it cleaner after recent wait-timeout changes.

**Comment by marcelo-gonzalez** [comment]
Ah i forgot about this! yeah i could give it a try :) i'll post some ideas this week

**Comment by marcelo-gonzalez** [comment]
Got a little busy with some other stuff last week, didn't get a chance to get to it. Looks like it's fixed by c2aea7b748d93 though (very simple fix hehe :) ). So i guess we can just close this issue then? Do you think it's still worth refactoring _io_uring_get_cqe() like you suggested before?

**Comment by isilence** [comment]
Ah, yes, it was getting in my way for fixing live-locks. Yes, it pretty is. Also, check out
https://lore.kernel.org/io-uring/c9550dcf-ce53-c214-8c4b-6165ad6605a9@gmail.com/
